### DISCUSSION TITLE
Sentinel-2-L2A xml metadata to a raster in R

### DISCUSSION THREAD

**Discussion by MullerMitro** [microsoft/PlanetaryComputer#155] [Q&A]
I have been using rstac and gdalcubes R packages to compute zonal statistics from time series of Sentinel-2-L2A data. I would like to be able to extract the sun zenith angle information at each location also. The Sentinel-2 STAC item collection includes asset named "granule-metadata". This data is stored in xml format. It seems to store gridded information on incidence angles with 5km spatial resolution (Geometric_Info$Tile_Angles$Sun_Angles_Grid$Zenith). 

I am searching for a  solution on how to parse the xml data and transform the gridded sun zenith angle values to a raster band.
Any suggestions how to solve this in R?

**Comment by gilbertocamara** [comment]
Dear Mitro

I am cc your message to Felipe Souza and Rolf Simoes, who are the developers and maintainers of “rstac”, and work in my team in Brazil. I hope they are able to help you.

All the best
Gilberto
============================
Prof Dr Gilberto Camara
Senior Researcher
National Institute for Space Research (INPE), Brazil
https://gilbertocamara.org/
=============================

**Reply by MullerMitro** [reply]
I appreciate your help. I look forward to hearing from Felipe Souza and/or Rolf Simoes. 

Best regards,
Mitro Müller

**Comment by TomAugspurger** [comment]
xref to https://github.com/stactools-packages/sentinel2/issues/50, which is requesting that this information be added to the stactools package, so that it's available through the STAC metadata.

**Comment by rolfsimoes** [comment]
I don't know any solution in R for transforming the gridded sun zenith and azimuth angle values to a raster. However, I have found a possible solution in the paper "Sentinel-2 MultiSpectral Instrument (MSI) data processing for aquatic science applications: Demonstrations and validations" by [Pahlevan et al. (2017)](https://doi.org/10.1016/j.rse.2017.08.033). The paper includes a Python script in the annex that can be used to compute the angles. 

The script, written in Python 2.7, implements an algorithm for determining the orbit and viewing angle observations of the Sentinel-2A sensor. It reads the values from the XML file that you have mentioned. The algorithm then computes the angles for each pixel from an orbit model.

To run the script, you must download a Sentinel-2 image in the SAFE format. Then, you should uncompress the file and specify the MTD_TL.xml file located in the granule folder as the script argument. Note that the script must also access the QI_DATA subfolder inside the SAFE structure.

The script can generate a raster of sun azimuth and zenith angles in HDR format, but you might be able to adapt this for your purposes.

You can find additional information in the following ESA forum issues:
* https://forum.step.esa.int/t/sentinel-2-viewing-angles-interpolation-on-per-pixel-basis/2776/4
* https://forum.step.esa.int/t/generate-view-angles-from-metadata-sentinel-2/5598/2

@oldlipe, do you know any other solution?

**Reply by MullerMitro** [reply]
Thank you for your input. 

My search on ready R solutions ended empty-handed. I was thinking something in lines of parsing the xml data to lists, converting lists to arrays and converting these to suitable raster format. Then let gdalcubes worry about the interpolation when creating a data cube.

I will look at the python script noted for inspiration. Thanks for the tip!

Currently I am still working on set of defined plots, so I can just calculate the sun zenith angle for each plot center using x, y, date and S2 acquisition time. This buys me some time before moving into larger scale processing.

**Comment by safalabolo** [comment]
Dear all,

https://drive.google.com/drive/folders/1g_hG5LB1HeHEA2_lz1WDSDlnbQT0npLK?usp=drive_link
you will find the main R file that loads the libraries and functions.
These are not scripts written by me; I simply assembled them following the instructions provided here:
https://github.com/ranghetti/sen2r/blob/main/R/s2_angles.R
and retrieved the necessary files from this section:
https://github.com/ranghetti/sen2r/tree/main/R

I had to make some modifications, for example, the file s2_angles.R became s2_angles_rev.R due to an intervention at line 146.

There is no full agreement between the angles derived using these scripts and those derived from SNAP (see scatterplots for reference), but this may be due to simplifications introduced in the scripts to produce the SAA and OAA angles, as well as the 500x500 m grid adopted in SNAP.

Kind regard
Safalabolo
